main.go

```diff
@@ -113,3 +112,126 @@ func main1() int {
+//
+// Then $moddir will be mounted as /start and the working directory will be
+// /start/cmd/blah.
+//
+// GOPATH mode (to be implemented)
+// -------------------------------
+// Assuming:
+//
+// * GOPATH=/gp1:/gp2
+// * that the package github.com/a/b/cmd/blah exists within /gp1/src
+// * a working directory of /gp1/src/github.com/a/b
+// * that we run go test -exec='...' ./cmd/blah
+//
+// Then /gp2 will be mounted as /gopath2, /gp1 will be mounted as /gopath1,
+// GOPATH=/gopath1:/gopath2 will be set, and the working directory will be
+// /gopath2/src/github.com/a/b
+//
+// TODO: implement GOPATH logic; for now we assume ad hoc mode logic.
+//
+// Ad hoc mode
+// -----------
+// Assuming:
+//
+// * a working directory of $dir
+//
+// Then $dir will be mounted as /start and the working directory will be
+// /start
+func buildDockerFlags() ([]string, error) {
+	var res []string
+
+	var env struct {
+		GOCACHE string
+		GOPATH  string
+		GOMOD   string
+	}
+	envCmd := exec.Command("go", "env", "-json")
+	out, err := envCmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to run %v: %v\n%s", strings.Join(envCmd.Args, " "), err, out)
+	}
+	if err := json.Unmarshal(out, &env); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal %v output: %v", strings.Join(envCmd.Args, " "), err)
+	}
+
+	// Normalise GOPATH elements for symlinks for the purposes of
+	// GOPATH mode below
+	var gp []string
+	var dockerGp []string // the gopath elements for the container
+	for i, v := range strings.Split(env.GOPATH, string(os.PathListSeparator)) {
+		ev, err := filepath.EvalSymlinks(v)
+		if err != nil {
+			return nil, fmt.Errorf("failed to filepath.EvalSymlinks(%q)", v)
+		}
+		gp = append(gp, ev)
+		dv := fmt.Sprintf("/gopath%v", i+1)
+		res = append(res, fmt.Sprintf("--volume=%v:%v", ev, dv))
+		dockerGp = append(dockerGp, dv)
+	}
+	res = append(res,
+		"--env=GOPATH="+strings.Join(dockerGp, string(os.PathListSeparator)),
+		fmt.Sprintf("--volume=%v:/gocache", env.GOCACHE),
+		"--env=GOCACHE=/gocache",
+	)
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get working directory: %v", err)
+	}
+
+	if env.GOMOD != "" && env.GOMOD != os.DevNull {
+		// we are in module-aware mode and have a main module
+		var mod struct {
+			Path string
+			Dir  string
+		}
+		modCmd := exec.Command("go", "list", "-m", "-json")
+		out, err := modCmd.CombinedOutput()
+		if err != nil {
+			return nil, fmt.Errorf("failed to run %v: %v\n%s", strings.Join(modCmd.Args, " "), err, out)
+		}
+		if err := json.Unmarshal(out, &mod); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal %v output: %v", strings.Join(modCmd.Args, " "), err)
+		}
+		rel, err := filepath.Rel(mod.Dir, wd)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine %v relative to %v: %v", wd, mod.Dir, err)
+		}
+		res = append(res,
+			fmt.Sprintf("--volume=%v:/start", mod.Dir),
+			fmt.Sprintf("--workdir=%v", filepath.Join("/start", rel)),
```

Comment:
nit: we use unix-y paths like `/start` and `/init`, so for consistency, this should be `"/start/"+rel`.

or if you really want, path.Join, but not filepath.Join.

Also because the file path separator of the host machine doesn't need to be the same as that of the target docker image. Imagine dockexec on Windows running a Linux docker image under a VM.

(I get that you use filepath.Rel above, but there is no path.Rel, so I think that's fine)

Reply:
Good catch, thanks. The work for calculating rel is always going to be on the host, so that's correct to my mind. But the path construction for the volume target is indeed going to need to be guest-specific. Will fix this to `path.Join` and add a comment... because there's going to be plenty of places we need to fix this, and/or extract the calculation logic to a separate function.

Reply:
I think it's fine if we assume the target to always be unix for now, but then we should keep the code simple.

Once we want to get Windows images to work properly, then we'll have a ton of work to do. Not only to replace path separators, but also to specify the drive like `C:`.